wordpress-mobile/WordPress-iOS
# Blogging Prompts: Fix Core Data concurrency issues in scheduler

Fixes #20842

> **Warning**:
> Please note that this targets the release/22.6 branch, which is already in code freeze.

## Description

When scheduling or rescheduling prompts, the related methods make a call to `UNNotificationCenter` to either request for push authorization or fetch notification settings. However, the notification center does not guarantee that the completion block will be executed in the caller's queue. Since the completion block code involves Core Data, this raises a multithreading violation when run in debug mode.

This PR fixes the issue by wrapping the Core Data-related code within these methods in the managed object context's queue.

Additionally, this also improves the thread-safety of `BloggingPromptsService`'s init method by ensuring that any access to Core Data properties is performed within that object's context queue.

## To test

- From Xcode, launch the Jetpack app with `-com.apple.CoreData.ConcurrencyDebug 1` launch argument.
- Log in with your WordPress.com account.
- Go to Site Menu > Site Settings.
- Set Blogging Reminders with prompts enabled.
- 🔎 Verify that the reminders are successfully set and that no multithreading violation is raised.
- Relaunch the Jetpack app from Xcode.
- 🔎 Verify that no multithreading violation is raised.
- Go to Site Menu > Site Settings.
- In the Blogging Reminders setting, uncheck all the days from the reminders.
- 🔎 Verify that the reminders are successfully set and that no multithreading violation is raised.

Additionally, please also smoke test Blogging Reminders — creating, updating, or removing schedules; with and without prompts.

## Regression Notes
1. Potential unintended areas of impact
Should be none. The logic is kept as close to the original as possible to minimize impact.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes, and the existing unit tests are passing.

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:


UI Changes testing checklist:

N/A — this PR doesn't contain UI-related changes.
